### PR TITLE
[WIP] Refactoring cache to support PSR-6 and Symfony cache natively

### PR DESCRIPTION
**Pull request by moufmouf**

### CODE REVIEW

**Comment by oojacoboo** [comment]
@moufmouf Can this be closed?  Does this PR that was merged in for 4.3.0 address this issue?

https://github.com/thecodingmachine/graphqlite/pull/373

**Comment by TamasSzigeti** [comment]
I feel we could do more to clean up caching. 
The most elegant would be to just use PSR interfaces everywhere, and let the user decide if they go with symfony or something else.
There should also be a cleaner separation re what kind of caching do we want in each place, following system cache vs app cache distinction in symfony (basically file/apcu vs redis or sthg like that), especially if people are using it outside of symfony framework. We probably only want system cache, but currently it's hard to follow what is being injected or used.

**Comment by oojacoboo** [comment]
@TamasSzigeti I entirely agree.  I'm in support of sticking with PSR interfaces for everything and letting the `SchemaFactory` handle the assignment.

**Comment by moufmouf** [comment]
Hey guys!

I took the route of providing a default implementation because I fear that users might use a wrong type of cache by default.

Let me explain:

Most of the cache is used to optimize the annotation parsing. Annotation parsing happens at the server level. So it makes a lot of sense to use a "local" cache that is typically quite fast (either by using APCu or a PHP opcode based cache).
Using a remote cache (like Redis or Memcache, or a database backed cache) would tremendously slow down the everything.

So taking the default cache set up in Laravel or Symfony would be a mistake (because on many systems, the default cache is Redis). The user would need to configure a special cache for GraphQLite anyway, and if this is not very properly documented, it is likely the user will make a mistake that will make GraphQLite very slow.

Since in my view, there are only 2 types of caches that are viable for the annotations (i.e. APCu or Opcode/PHPFiles), I said to myself it would be better to make this choice by default for the end-user (even if we still allow overriding)

Don't get me wrong, I'm a huge proponent of PSR's (I'm the PSR-11 co-editor). But I also believe good defaults are important.

**Comment by TamasSzigeti** [comment]
Yes, that was indeed my point too above, but I was thinking of just documenting it. So if you feel like directly depending on symfony cache is fine, then it's fine for me, actually even easier to implement. Still some work to make it all consistent though, and it's not entirely clear for me what this PR is doing. Symfony cache has psr converters back and forth

**Comment by oojacoboo** [comment]
It could also be made more obvious by using suggestive method names in the `SchemaFactory`.  Something like...

- `setLocalCache`
- `setAnnotationCache`

**Comment by hjerichen** [comment]
Stupid Question, but is an annotation cache still needed when dropping annotations in favor of attributes in PHP 8?
If not and dropping PHP 7 support is not far away, then this discussion may not be needed.

**Comment by oojacoboo** [comment]
@hjerichen I don't think support for PHP 7 should be dropped anytime soon, but you do bring up a valid point and consideration.

**Comment by oojacoboo** [comment]
Any objections to closing this PR in favor of planning through #417 and the improvements already provided in #373?
